lightning/src/routing/router.rs

```diff
@@ -333,84 +592,262 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
-			add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, ChannelFeatures::empty(), 0);
+		// Step (1).
+		// If a caller provided us with last hops, add them to routing targets. Since this happens
+		// earlier than general path finding, they will be somewhat prioritized, although currently
+		// it matters only if the fees are exactly the same.
+		for hop in last_hops.iter() {
+			let have_hop_src_in_graph =
+				if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat)) = first_hop_targets.get(&hop.src_node_id) {
+					// If this hop connects to a node with which we have a direct channel, ignore
+					// the network graph and add both the hop and our direct channel to
+					// the candidate set.
+					//
+					// Currently there are no channel-context features defined, so we are a
+					// bit lazy here. In the future, we should pull them out via our
+					// ChannelManager, but there's no reason to waste the space until we
+					// need them.
+					add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features.to_context(), 0, recommended_value_msat);
+					true
+				} else {
+					// In any other case, only add the hop if the source is in the regular network
+					// graph:
+					network.get_nodes().get(&hop.src_node_id).is_some()
+				};
+			if have_hop_src_in_graph {
+				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+				// really sucks, cause we're gonna need that eventually.
+				let last_hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+					Some(htlc_minimum_msat) => htlc_minimum_msat,
+					None => 0
+				};
+				let directional_info = DummyDirectionalChannelInfo {
+					cltv_expiry_delta: hop.cltv_expiry_delta as u32,
+					htlc_minimum_msat: last_hop_htlc_minimum_msat,
+					htlc_maximum_msat: hop.htlc_maximum_msat,
+					fees: hop.fees,
+				};
+				add_entry!(hop.short_channel_id, hop.src_node_id, payee, directional_info, None::<u64>, ChannelFeatures::empty(), 0, recommended_value_msat);
+			}
 		}
-	}
 
-	while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
-		if pubkey == *our_node_id {
-			let mut res = vec!(dist.remove(&our_node_id).unwrap().3);
-			loop {
-				if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
-					res.last_mut().unwrap().node_features = features.to_context();
-				} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
-					if let Some(node_info) = node.announcement_info.as_ref() {
-						res.last_mut().unwrap().node_features = node_info.features.clone();
+		// At this point, targets are filled with the data from first and
+		// last hops communicated by the caller, and the payment receiver.
+		let mut found_new_path = false;
+
+		// Step (2).
+		'path_construction: while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, value_contribution_msat, .. }) = targets.pop() {
```

Comment:
Can you comment the exit condition of leaving `path_construction` due to exhaustion of `targets`. "path_construction failure to generate a new path (`found_new_path`) means we don't have enough outbound liquidity on the known graph to route across" ?